### ISSUE TITLE
F401 unused import false positive

### ISSUE THREAD

**Issue by sbrugman** [charliermarsh/ruff#4656]
False negative
```
import importlib # this package is unused, but not detected
import importlib.util 

package = "ruff"
is_present = importlib.util.find_spec(package)
```
When `from importlib import util` is used, correct behaviour is observed
```
import importlib # this package is unused, and detected
from importlib import util 

package = "ruff"
is_present = util.find_spec(package)
```

**Comment by charliermarsh** [comment]
This is consistent with Pyflakes and I think it's intentional, but submodule imports are really complex, I'm not certain whether it's safe in the general case to mark `import a` as unused in:

```py
import a
import a.b

a.b.c()
```

It might be safe? I believe we had a PR long ago that addressed this: #58.

(Here's a good example of why / how submodules can be difficult, though it's not exactly the same as above: https://github.com/asottile/scratch/wiki/Why-doesn't-flake8-mark-this-as-an-unused-import%3F.)

**Comment by sbrugman** [comment]
Thanks for sharing the resource.

For what it's worth PyCharm does flag `import a` as unused import in the example above.

Can we think of examples where we would like to treat that example differently from this:

```python
import b
import a.b

a.b.c()
```
(The case in the linked resource above is the inverse)

**Comment by konstin** [comment]
I'm struggling to find a case where `import a.b` is semantically different from `import a; import a.b` given that `import a.b` also runs `a/__init__.py` (https://docs.python.org/3/reference/import.html#the-meta-path, fourth paragraph)

**Comment by charliermarsh** [comment]
Yeah, it might be the case that if we have `import a; import a.b` and then `a.b` is used but not _just_ `a`, then we can mark `import a` as unused.

On the other hand, per Anthony's linked note, if we have `import a; import a.b`, and then `a.z.x` is used, it seems like we can't actually treat `import a.b` as unused, since it _could_ modify `a`.

Similarly, if you have `import a; import a.b`, and then you do `print(a); print(a.b.c)`, I don't think we can treat either import as unused.

**Comment by pkulev** [comment]
> On the other hand, per Anthony's linked note, if we have import a; import a.b, and then a.z.x is used, it seems like we can't actually treat import a.b as unused, since it could modify a.

Oh, indeed. I think that if something happens on import time, like filling some global variables from other modules (handlers registration or some `__init_subclass__` registration magic) it must be visible, and `# noqa: import for side-effects` is a good indicator of such things. I would prefer throwing a bunch of noqa's for such rare cases, than considering it normal.
